[FEATURE] blast module & activation of demo builds & fixes to demos

h-2 opened the pull request:
see #938 for a description.


@rrahn + @esiragusa although there are still some TODOs, please comment!

h-2 commented:
The only big thing in here not blast is the #773, please review that first.

esiragusa commented on demos/blast/blast_in_example.cpp:
```diff
@@ -0,0 +1,43 @@
+
+    std::string inPath = std::string(SEQAN_PATH_TO_ROOT()) + "/tests/blast/plus_header_defaults.blast";
+
+    BlastTabularIn<TContext> in(toCString(inPath));
```
Should be called just BlastFileIn.

esiragusa replied:
You should catch the exceptions thrown by the constructor.

h-2 replied:
> Should be called just BlastFileIn.

There are two formats currently implemented: BlastTabular and BlastReport. I could rename the FormattedFiles from BlastTabularOut to BlastTabularFileOut a.s.o
Do you mean that?

> You should catch the exceptions thrown by the constructor.

Yes!

esiragusa replied:
There has to be one BlastFile which supports the two formats TabularBlast and ReportBlast. Like BamFile which supports Sam and Bam.

esiragusa commented on demos/blast/blast_in_example.cpp:
```diff
@@ -0,0 +1,43 @@
+    typedef Align<String<AminoAcid>, ArrayGaps> TAlign;
+    typedef BlastMatch<CharString, CharString, uint32_t, TAlign> TBlastMatch;
+    typedef BlastRecord<CharString, CharString, uint32_t, TAlign> TBlastRecord;
+    typedef BlastIOContext<Blosum62> TContext;
```
Do you have to know in advance the correct scoring scheme? Is Blosum62 reasonable as default?

h-2 replied:
> Do you have to know in advance the correct scoring scheme?

Yes, it is a required spec for the BlastScoringScheme which is in turn required for Blast IO.

> Is Blosum62 reasonable as default?

Since all BLAST programs other than BLASTN score in protein space and Blosum62 is the default there, I think it should be here, as well.

esiragusa commented on demos/blast/blast_in_example.cpp:
```diff
@@ -0,0 +1,43 @@
+
+using namespace seqan;
+
+int main()
```
If you add a tutorial in the manual, please reuse this demo - don't make a new one.

h-2 replied:
Yes, I will.

esiragusa commented on demos/blast/blast_in_example.cpp:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    return 0;
+}
```
Please add a new line at the end of all your files.

h-2 replied:
Ah, that one escaped me. Of course all files should end on on newline.

esiragusa commented on demos/blast/blast_in_example.cpp:
```diff
@@ -0,0 +1,43 @@
+int main()
+{
+    typedef Align<String<AminoAcid>, ArrayGaps> TAlign;
+    typedef BlastMatch<CharString, CharString, uint32_t, TAlign> TBlastMatch;
```
Please change all types from (u)intXX_t to __(u)intXX - see e.g. https://github.com/seqan/seqan/blob/master/include/seqan/platform/platform_gcc.h

h-2 replied:
I can change this for now, but I think we should discuss changing it instead in all other places to use the official typedef, since all of our target compilers now have this, I think.

esiragusa replied:
Yep we have them with C++11 (see http://www.cplusplus.com/reference/cstdint/). But please change it back for now.

esiragusa commented on demos/blast/blast_in_example.cpp:
```diff
@@ -0,0 +1,43 @@
+            std::cout << "E-Value of best HSP: " << front(record.matches).eValue << "\n";
+
+        // if there is anything unexpected, tell the user about it
+        if (length(context(in).conformancyErrors) > 0)
```
!empty(context(in).conformancyErrors)

h-2 replied:
+1

esiragusa commented on tests/blast/test_blast_statistics.h:
```diff
@@ -0,0 +1,280 @@
+    }
+
+    // Blosum45
+    {
```
You could have refactored this test...

h-2 replied:
But all ther numbers are actually different? How would you have refactored this?

esiragusa replied:
Use statically initialized vectors of double?

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+template <typename TNum,
+          typename std::enable_if<Is<NumberConcept<TNum>>::VALUE, int>::type = 0>
+inline void
+clear(TNum & num)
```
Why here in the test file?

h-2 replied:
Was a leftover from before that isn't needed anymore.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+template <typename TArg,
+          typename... TArgs>
+inline void
+clear(TArg & arg, TArgs & ... args)
```
As above.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+    clear(args...);
+}
+
+void _test_blast_read_tabular_match(std::string const & path,
```
We use camel case for functions...

h-2 replied:
Except for test names, you mean? ^^
But true, this is not a test but a function that is called by a test.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+
+    std::string inPath = std::string(SEQAN_PATH_TO_ROOT()) + std::string(path);
+
+    std::ifstream ifstream(toCString(inPath),
```
Please, all tests should test the public FormattedFile interface only.

h-2 replied:
Hm, I can change this, but then we loose a lot of currently tested functionality...

esiragusa replied:
Which functionality is outside of the FormattedFile interface? Is it relevant? This additional complexity blows up not only tests but also dox and manual.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+    }
+
+    // read header of file
+    readHeader(context, it, BlastTabular());
```
Why does readHeader() need a format tag?

h-2 replied:
Because its different for BlastTabular and BlastReport?

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+        exceptComment = e.what();
+    }
+
+    SEQAN_ASSERT(startsWith(exceptComment, "Unable to convert"));
```
Do not assert assertion comments.

h-2 replied:
Ok.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+    std::string inPath = std::string(SEQAN_PATH_TO_ROOT());
+    inPath += (custom ? PLUS_HEADER_CUSTOM : PLUS_HEADER_DEFAULTS);
+
+    std::ifstream ifstream(toCString(inPath),
```
As above, please test through the public FormattedFile interface only.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+    SEQAN_ASSERT(!onMatch(context, BlastTabular()));
+    SEQAN_ASSERT(onRecord(context, BlastTabular()));
+
+    readRecordHeader(r, it, context, BlastTabular());
```
A function called readRecordHeader() seems very confusing, as we have already readRecord() and readHeader(). Also, does it need to be public?

h-2 replied:
It was publicly documented before, so it is still public. I can change it to private, but I don't see whats wrong with the name. Every record has a header; since SeqAn defines the header to be file-global, I used RecordHeader to show that it is the header of the record. This is also explained in current dox.

esiragusa commented on tests/blast/test_blast_input.h:
```diff
@@ -0,0 +1,1166 @@
+    _test_blast_read_tabular_with_header(true);
+}
+
+SEQAN_DEFINE_TEST(test_blast_read_header_tabular_with_header_legacy)
```
Can't you refactor all these tests?

esiragusa replied:
Please remember to:
* Test only public interfaces
* Stick to **small and generic** unit tests

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+                    int const,
+                    BlastReport const & /**/)
+{
+    if (format < 3)
```
What is format < 3? Do you mean <3? :trollface:

h-2 replied:
;)
It is an internal variable that is used to re-use code for different tests. If I remove everything except the FormattedFile interface than this will go away.

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+    if (format < 3)
+    {
+        writeHeader(file, context, BlastReport());
+        for (auto const & r : records)
```
Don't we have writeRecord**s**()?

h-2 replied:
Only SeqFileOut has this I think. At least it is not in the dox for other formats...

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+        writeFooter(file, context, BlastReport());
+    } else
+    {
+        BlastReportOut<BlastIOContext<TScore, TString, p, h>> out(context, file, BlastReport());
```
Should be called BlastFileOut.

h-2 replied:
Not really. It has to be templatized somewhere. Only templatizing the context at least moves the variability to one place.
Since downstream functions don't work on the FormattedFile but on the context and the iterator, obviously we can't move sepcializations from the context to the FormattedFile.

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+                    int const custom,
+                    BlastTabular const & /**/)
+{
+    switch (format)
```
:bomb:

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+
+                for (auto const & m : r.matches)
+                {
+                    if (custom <= 1)
```
:spaghetti:

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+    String<TBlastRecord> records;
+    resize(records, 3);
+
+    context.dbName = "The Foo Database";
```
Note that other FormattedFileContextes (e.g. http://docs.seqan.de/seqan/develop/class_BamIOContext.html, http://docs.seqan.de/seqan/develop/class_VcfIOContext.html) already have standardized accessors for these properties. Please stick to that interface.

h-2 replied:
Then we need to discuss these accessors because obviously they were not intended to be general. I don't have contigs, and I can't expect users to know that they have to say "contig" when they mean "protein in protein database".
[But I do agree that have the same accessors is very much desirable]

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+    SEQAN_ASSERT(isValid(context.scoringScheme));
+
+    String<TBlastRecord> records;
+    resize(records, 3);
```
FYI this will call the Generous resize.

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+}
+
+template <typename TScore, typename TString, BlastProgram p, BlastTabularSpec h>
+void test_blast_write_tabular_impl(int const format, // 0 only matches; 1 matches+headers; 2 records; 3 formattedFile
```
Please don't use these numbers.

h-2 replied:
see above, it will go away if we only test formattedfile.

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+    versionLine.append(", http://www.seqan.de]\n");
+
+    std::string fieldsLine = "# Fields: ";
+    if (context.legacyFormat)
```
The following 100 lines of code are hard to understand... :spaghetti:

h-2 replied:
Well, its the construction of the expected output. It can be done with less :spaghetti: but it will mean lots of duplication.

esiragusa commented on tests/blast/test_blast_output.h:
```diff
@@ -0,0 +1,814 @@
+
+// TEST MATCH, WITHOUT HEADER, DEFAULTS FIELDS
+
+SEQAN_DEFINE_TEST(test_blast_write_match_tabular_run_time_context_args)
```
We have a generic testing framework. The idea was to avoid repeating tests like in the following 200 lines.

h-2 replied:
I haven't done testing on this scale before, can you point me to the relevant documents describing how to do this in a nicer way?

esiragusa commented on include/seqan/blast/blast_tabular_out.h:
```diff
@@ -0,0 +1,799 @@
+            _untranslateQPositions(effectiveQStart, effectiveQEnd, match.qFrameShift, match.qLength,
+                                   context.blastProgram, BlastProgramTag<p>());
+            write(s, effectiveQStart);
+        } break;
```
Breaks go on a new line.

h-2 replied:
ok.

esiragusa commented on include/seqan/blast/blast_tabular_out.h:
```diff
@@ -0,0 +1,799 @@
+//         case ENUM::S_SEQ: write(s,  * ); break;
+        case BlastMatchField<>::Enum::E_VALUE:
+        {
+            std::string formatString;
```
Please (re)use a buffer in the context.

h-2 replied:
For strings this short this is guaranteed to be storage from the stack; accessing a potentially larger buffer and overwriting it could actually be slower.
But if we have a policy for this I can change it of course.

esiragusa commented on include/seqan/blast/blast_tabular_out.h:
```diff
@@ -0,0 +1,799 @@
+        } break;
+        case BlastMatchField<>::Enum::BIT_SCORE:
+        {
+            std::string formatString;
```
See above.

esiragusa commented on include/seqan/blast/blast_tabular.h:
```diff
@@ -0,0 +1,661 @@
+     * @brief An array of bools revealing whether the Blast I/O module supports printing this field
+     */
+    //TODO(c++14): change to std::bitset that is initialized with binary literal
+    static constexpr bool const implemented [] =
```
I think there is a false missing at line 610 :trollface:

h-2 replied:
?

esiragusa commented on include/seqan/blast/blast_tabular.h:
```diff
@@ -0,0 +1,661 @@
+     * @brief An array of CStrings representing the <b>column label</b> of each possible field; for the
+     * @link BlastIOContext::legacyFormat @endlink, use @link BlastMatchField::legacyColumnLabels @endlink instead.
+     */
+    static constexpr char const * const columnLabels [] =
```
Why constexpr??

h-2 replied:
Because I can :gangnamstyle:

esiragusa commented on include/seqan/blast/blast_tabular.h:
```diff
@@ -0,0 +1,661 @@
+
+
+template <typename TVoidSpec = void>
+struct BlastMatchField
```
I think we should discuss together how to represent this stuff in a better way.

h-2 replied:
Hm, sure.

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+ * @val BlastProgram BlastProgram::UNKNOWN
+ * @brief Unkown type. Used internally and to signify that the type could not be inferred from the file.
+ */
+enum class BlastProgram : uint8_t
```
These should have been file format tags.

h-2 replied:
As I previously said: before there was one BlastFormat which had multiple template arguments for the fileformate (Tabular, TabularWithHeader, Report, poossibly XML...), the program type (BlastN, BlastP...), the generation (blastall vs blast+). You said that it was super confusing and that it needed to change.
Then I treied it with the TagSelector but obiously this then results in huge confusing number of possible combinations that all get a tag... so instead I want with the above :)

esiragusa replied:
I mean compile time format traits...

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+ */
+
+// this is always evaluated at compile-time
+template <BlastProgram p>
```
Template argument must be prefixed by T, e.g. TBlastProgram.

h-2 replied:
But only if its a type, or not? This is a value.

esiragusa replied:
Ah... did you (kind of) reimplement the tag selector for enums?

h-2 replied:
Yes, sort of. But in my case there are not ever any needless checks, if values are selected at compile-time. And even if they are dynamic, there is no recursion (which is bad because it can't be inlined and its also bad for branch prediction I think).

Seeing that there are 5 blast program formats, 3 implemented file types and legacy and plus formats for each there would already be 30 combinations of fileformats to check whenever a record a is written. This is totally not feasible.

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+
+// this will likely be evaluated at run-time
+constexpr BlastProgram
+getBlastProgram(BlastProgram const p, BlastProgramTag<BlastProgram::UNKNOWN> const &)
```
What's different from the above function?

h-2 replied:
one returns left arg, the other returns the right arg?

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+ */
+
+constexpr bool
+qHasRevComp(BlastProgram const p)
```
These should have been metafunctions.

h-2 replied:
Maybe we should discuss this. As we are targetting c++11 anyway, I don't see think it is better do make things needlessly more complex (however I do see value in consistency).

h-2 replied:
> These should have been metafunctions.

They are used both as regular and as constexpr functions, so any other solution would have meant more code, less clarity :)

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+constexpr bool
+qHasRevComp(BlastProgram const p)
+{
+    return ((p==BlastProgram::BLASTP) || (p==BlastProgram::TBLASTN))
```
return !(p == BlastProgram::BLASTP || p == BlastProgram::TBLASTN)

h-2 replied:
nice catch!

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+ */
+
+constexpr bool
+qIsTranslated(BlastProgram const p)
```
Previous comments apply to all functions below. In addition, the code below is extremely repetitive...

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+inline std::string const
+_programTagToString(BlastProgram const _p)
+{
+    switch(_p)
```
What about mapping the types to a static const char array?

h-2 replied:
You are right, this would be a much better solutation. I will change this.

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+}
+
+// ----------------------------------------------------------------------------
+// _programTagToString()
```
In general, I don't see what you are trying to do here with _programTagToString() and why it is so complicated...

h-2 replied:
Yes, see above.

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+inline BlastProgram
+_programStringToTag(TString const & str)
+{
+    if (str == _programTagToString<BlastProgram::BLASTN>())
```
What about searching the string back in the static const char array?

h-2 replied:
Yes, see above.

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+inline typename Size<TString>::Type
+_firstOcc(TString const & str, typename Value<TString>::Type const & val)
+{
+    typedef typename Size<TString>::Type S;
```
Typenames start with T

h-2 replied:
Yes.

esiragusa commented on include/seqan/blast/blast_base.h:
```diff
@@ -0,0 +1,669 @@
+
+template <typename TString>
+inline typename Size<TString>::Type
+_firstOcc(TString const & str, typename Value<TString>::Type const & val)
```
Isn't this std::find?

h-2 replied:
True, I will replace it.

esiragusa commented on include/seqan/blast/blast_io_context.h:
```diff
@@ -0,0 +1,383 @@
+ */
+
+template <typename TScore_ = Blosum62,
+          typename TString_ = std::string,
```
Is it important to change TString?

h-2 replied:
Not important. Using std::string instead of CharString is a little more efficient in some places, but I though maybe someone would want to use CharString for some reasons.

esiragusa replied:
Then let's remove it from the main template parameters, one could still tweak it through Fibre/Member, or one last optional config/traits object.

esiragusa commented on include/seqan/blast/blast_io_context.h:
```diff
@@ -0,0 +1,383 @@
+
+template <typename TScore_ = Blosum62,
+          typename TString_ = std::string,
+          BlastProgram _p = BlastProgram::UNKNOWN,
```
These should have been file format tags.

h-2 replied:
see above

esiragusa commented on include/seqan/blast/blast_io_context.h:
```diff
@@ -0,0 +1,383 @@
+        if (!legacyFormat)
+            append(versionString, "+");
+        append(versionString, " [I/O Module of SeqAn-");
+        append(versionString, std::to_string(SEQAN_VERSION_MAJOR));
```
This doesn't work if TString is seqan::CharString.

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+ * @implements LessThanComparableConcept
+ * @headerfile <seqan/blast.h>
+ * @signature struct BlastMatch<TQId, TSId, TPos, TAlign> { ... };
+ * @brief An data structure to hold a blast match, also known as high-scoring segment pair (HSP)
```
Why don't you call it HSP then?

h-2 replied:
I thought "match" is a more general term, as it used in other palces in SeqAn, too.

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+ *
+ */
+
+template <typename TQId = std::string,
```
Do we need all these template parameters? Maybe only the ArrayGaps spec?

h-2 replied:
In some situations you will want to change the types, e.g. when you have all IDs in a ConcatDirect StringSet, it makes much more sense to also use Infixes for the IDs, so you don't need to copy.

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+     * @var TSId BlastMatch::sId;
+     * @brief The verbose Id of the subject.
+     */
+    TQId            qId;
```
Why don't you factorize out query and subject information? You repeat id, start, end, length, frameshift for both...

h-2 replied:
I thought it would be more confusing to add another layer of members. And I am not sure it would actually reduce the lines of code or increase its maintainability. But I can change it, sure!

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+        return (bitScore >= bm2.bitScore);
+    }
+
+    inline void _clear()
```
Why do you define it as a member function?

h-2 replied:
could be changed.

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+     * @var std::list<TBlastMatch> BlastRecord::matches;
+     * @brief list of the contained matches
+     */
+    std::list<TBlastMatch>  matches;
```
Really a std::list? :scream:

h-2 replied:
Well the BlastMatches are rather fat objects and usually not too many. And  you always want to sort your matches before printing;my tests showed that this is measurably faster than with vectors.

esiragusa replied:
But why do you need a list rather than a vector? Do you insert elements in the middle?

h-2 replied:
> Do you insert elements in the middle?

No, but swapping is instant in a linked list, becaus only the pointers to the elements are changed. So for fat obejcts this is better than swapping in a vector.

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+inline SEQAN_FUNC_ENABLE_IF(Is<NumberConcept<TNumber> >, bool)
+_memberIsSet(TNumber const & in)
+{
+    return (in != std::numeric_limits<TNumber>::max());
```
FYI you don't need parentheses here nor above...

h-2 replied:
but it makes for nicer readability, or not?

esiragusa commented on include/seqan/blast/blast_record.h:
```diff
@@ -0,0 +1,384 @@
+ * <tt>Align<CharString, ArrayGaps></tt>
+ */
+
+template <typename TQId = std::string,
```
The same here for these template parameters.

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+ */
+
+template <typename TBlastIOContext = BlastIOContext<> >
+using BlastReportOut = FormattedFile<BlastReport, Output, TBlastIOContext>;
```
This will become a simple typedef when the BlastIOContext goes away.

esiragusa replied:
Should be called BlastFileOut.

h-2 replied:
> This will become a simple typedef when the BlastIOContext goes away.

I don't see any way of doing this without templatizing the final FormattedFile.

esiragusa replied:
Both score matrix and blast program type have to be format traits rather than io context template parameters!

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+
+template <typename T>
+struct MagicHeader<BlastReport, T> :
+    public MagicHeader<Nothing, T> {};
```
Really? Isn't it a string like BLAST*?

esiragusa replied:
FYI the public keyword is superfluous when you inherit a struct.

h-2 replied:
> Really? Isn't it a string like BLAST*?

It is '> BLAST*' or '> TBLAST*' for tabular with header, it is not defined for tabular without header, it  is 'BLAST*' or 'TBLAST*' for report.

> FYI the public keyword is superfluous when you inherit a struct.

Ah, I didn't know, but it makes sense of course.

h-2 replied:
In any case, since the magic header is always ambiguous, I will leave it as "nothing".

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+template <typename TSpec>
+inline bool guessFormat(FormattedFile<BlastReport, Output, TSpec> &)
+{
+    return true;
```
So there's no format guessing?

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+// ----------------------------------------------------------------------------
+
+constexpr const char *
+_blastReference()
```
Why don't you make them static const strings then?

h-2 replied:
It is the same really, the only difference is that to prevent global inclusion you would have to make the static const cstring a member of a templated struct (which is a hack by itself).

esiragusa replied:
Global inclusion into what?

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+                 TMatch const & m,
+                 BlastReport const &)
+{
+    char buffer[512] = "";
```
Why do you do = "" ?

h-2 replied:
I think I did something else before and need it is 0-initialized... it can go now.

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+{
+    char buffer[512] = "";
+
+    sprintf(buffer," Score =  %.1f bits (%d), Expect =  %.1g\n"
```
Why do you use a buffer + sprintf()?

h-2 replied:
Because it would take 3 pages of code to do the same by writing FormattedNumber :o

esiragusa replied:
Then we should come up with something better...

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+    while (aPos < m.alignStats.alignmentLength)
+    {
+        // Query line
+        sprintf(buffer, "Query  %-*d  ", numberWidth, qPos + effQStart);
```
Again, why do you use a buffer + sprintf()?

h-2 replied:
I have removed this in all places, except the one above, because this behaviour cannot be achieved with our `FormattedNumber`.

esiragusa commented on include/seqan/blast/blast_report_out.h:
```diff
@@ -0,0 +1,893 @@
+    else if (length(m.sId) < 66) // needs to be padded with ' '
+    {
+        write(stream, m.sId);
+        for (unsigned char i = 0; i < 66 -length(m.sId); ++i)
```
Use simply unsigned.

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+    typedef uint8_t size_t;
+
+    /* statics */
+    static constexpr size_t nParams = 8;
```
Why not just const??

h-2 replied:
Well anything that could be, should be constexpr. If all our types had proper constructors (and the compiler supports c++14), it would be possible to select the KarlinAltschulValues for a given scoring scheme at compile time :)

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+template <typename TSpec>
+struct KarlinAltschulValues<Pam250, TSpec>
+{
+    typedef uint8_t size_t;
```
TSize?

h-2 replied:
yes

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+
+    /* related to ScoringScheme conversion */
+    TValue _gapOpenBlast = 0;
+    TScore _internalScheme;
```
Shouldn't there be a constructor to pass a scoring scheme?

h-2 replied:
Because of the possibly confusing conversion of open scores I would rather have the people use the set*() functions instead.

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+struct BlastScoringScheme
+{
+    typedef typename Value<TScore>::Type TValue;
+    typedef typename KarlinAltschulValues<TScore, void>::size_t TNumValues;
```
Use Size/TSize.

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+    typedef typename KarlinAltschulValues<TScore, void>::size_t TNumValues;
+
+    /* related to ScoringScheme conversion */
+    TValue _gapOpenBlast = 0;
```
Why is it different from the base scoring scheme gap open?

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+    TScore _internalScheme;
+
+    /* parameter selection */
+    TNumValues parameterIndex = std::numeric_limits<TNumValues>::max();
```
What's that?

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+
+template <typename TMatrixSpec>
+inline double
+getLambda(BlastScoringScheme<Score<int, ScoreMatrix<AminoAcid, TMatrixSpec>>> const & scoringScheme)
```
Why do you have **two** versions of all these getters for ScoreMatrix and Simple?

h-2 replied:
Because the values are in different positions in the matrices.

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+// ----------------------------------------------------------------------------
+
+// Computes the length adjustment for E-value computation
+// Based on the public domain NCBI BLAST code by Tom Madden.
```
Please add the original license statement.

h-2 replied:
I have added a more prominent notice at the top of the file, but a public domain declaration implies that I don't have to do anything, it is not a license and comes with no terms.

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+
+template <typename TScore>
+inline TScore const &
+seqanScheme(BlastScoringScheme<TScore> const & bscheme)
```
getScoringScheme()?

h-2 replied:
I am using the term BlastScoringScheme and SeqanScoringScheme to differentiate between the behaviours, so it has to be clear that this function returns an object with SeqAn behaviour.

esiragusa commented on include/seqan/blast/blast_statistics.h:
```diff
@@ -0,0 +1,1076 @@
+{
+    uint64_t ql = match.qLength; // convert to 64bit, once
+    // length adjustment not yet computed
+    if (context._cachedLengthAdjustments.find(ql) == context._cachedLengthAdjustments.end())
```
Is ql the key or the value? Before you call container.find(ql) and then contaner[ql]...

esiragusa commented on include/seqan/blast/blast_tabular.h:
```diff
@@ -0,0 +1,661 @@
+{
+    static constexpr char const * VALUE[5] =
+    {
+        ".blast",
```
Hmm are we using the same set of extensions for different formats?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+          BlastProgram p,
+          BlastTabularSpec h>
+inline bool
+onRecord(BlastIOContext<TScore, TString, p, h> & context,
```
Private?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+          BlastProgram p,
+          BlastTabularSpec h>
+inline bool
+onMatch(BlastIOContext<TScore, TString, p, h> & context,
```
Private?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+
+template <typename TContext>
+inline bool
+onRecord(BlastTabularIn<TContext> & formattedFile)
```
Private?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+          BlastProgram p,
+          BlastTabularSpec h>
+inline void
+goNextLine(BlastIOContext<TScore, TString, p, h> & context,
```
Private?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+
+    do
+    {
+        if (std::regex_search(seqan::begin(context._lineBuffer),
```
Do you read/buffer the file line by line? Seriously? :scream:

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+          BlastProgram p,
+          BlastTabularSpec h>
+inline void
+skipRecordHeader(TFwdIterator & iter,
```
Private?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+{
+    if (context.legacyFormat)
+    {
+        if (SEQAN_ENABLE_DEBUG)
```
Why isn't this a preprocessor directive?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+          BlastProgram p,
+          BlastTabularSpec h>
+inline void
+skipMatch(TFwdIterator & iter,
```
What's the purpose of this function?

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+          BlastProgram p,
+          BlastTabularSpec h>
+inline void
+readRecordHeader(BlastRecord<TQId, TSId, TPos, TAlign> & r,
```
This function name is confusing.

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+
+    clear(blastRecord);
+
+    auto it = seqan::begin(context._lineBuffer); // move into line below when seqan supports && properly
```
Why do you use explicitly the namespace seqan? The iter tag is missing.

esiragusa commented on include/seqan/blast/blast_tabular_in.h:
```diff
@@ -0,0 +1,1295 @@
+            break;
+    }
+
+    if (length(blastRecord.matches) == 0)
```
Maybe empty() ?

h-2 commented:
All the warning that now appear in cash are related to demos. AFAICT the demos where not built automatically before at all, or we would have seen these warning already. I will fix what I can and include it in the pull request.

h-2 commented:
@esiragusa I have fixed lots of stuff and removed comments for issues that we discussed f2f. Can you have a look at the remaining points? Can you also look at the tests again, I have refactored them to use the FormattedFile interface only (except for the two low-level tests).
Maybe this can still be done more generic, but I didn't completely understand it from the sequence_v2 tests.

h-2 commented:
As far as I am concerned this is final, now that the very extensive tutorial is done, as well. 
@esiragusa please comment on any open issues, I want to merge this ASAP. Thanks!

h-2 commented:
If @esiragusa  is busy, can @rrahn have a look? I really want to merge this today so I have tomorrow to fix possible issues discovered during tonights' runs.
